apiserver/user/invitation_redeem.go

```diff
@@ -17,96 +18,91 @@ import (
-	return &userv1.RedeemOptions{}, false, ""
+func (ir invitationRedeemer) New() runtime.Object {
+	return &userv1.InvitationRedeemRequest{}
 }
 
-// Connect implements the REDEEM method for invitations.
-// It is used to redeem an invitation by a user.
+func (ir invitationRedeemer) Destroy() {}
+
+// Create implements redeeming invitations, it accepts `InvitationRedeemRequest`.
 // The user is identified by the username in the request context.
-// The token is taken from the path.
 // If the invitation is valid, the invitation is marked as redeemed, the user, and a snapshot of the invitations's targets are stored in the status.
 // The snapshot is later used in a controller to add the user to the targets in an idempotent and retryable way.
 // If user or token are invalid, the request is rejected with a 403.
-func (s *invitationRedeemer) Connect(ctx context.Context, name string, options runtime.Object, responder rest.Responder) (http.Handler, error) {
-	l := klog.FromContext(ctx).WithName("InvitationRedeemer.Connect").WithValues("invitation", name)
-	opts := options.(*userv1.RedeemOptions)
-	// Might come from the path, so we need to trim the leading slash
-	token := strings.TrimLeft(opts.Token, "/")
-
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		inv := &userv1.Invitation{}
-		if err := s.client.Get(ctx, client.ObjectKey{Name: name}, inv); err != nil {
-			responder.Error(err)
-			return
-		}
+func (s *invitationRedeemer) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, opts *metav1.CreateOptions) (runtime.Object, error) {
+	irr, ok := obj.(*userv1.InvitationRedeemRequest)
+	if !ok {
+		return nil, fmt.Errorf("not an InvitationRedeemRequest: %#v", obj)
+	}
 
-		tokenValid := inv.Status.Token != "" && inv.Status.ValidUntil.After(time.Now())
-		if inv.IsRedeemed() || !tokenValid || inv.Status.Token != token {
-			l.Info("invalid token")
-			forbidden(responder)
-			return
-		}
+	name := irr.Name
+	token := irr.Token
 
-		user, ok := userFrom(ctx, s.usernamePrefix)
-		if !ok {
-			l.Info("no allowed user found in request context", "usernamePrefix", s.usernamePrefix)
-			forbidden(responder)
-			return
-		}
+	l := klog.FromContext(ctx).WithName("InvitationRedeemer.Create").WithValues("invitation", name)
 
-		ts := make([]userv1.TargetStatus, len(inv.Spec.TargetRefs))
-		for i, target := range inv.Spec.TargetRefs {
-			ts[i] = userv1.TargetStatus{
-				TargetRef: target,
-				Condition: metav1.Condition{
-					Type:   userv1.ConditionRedeemed,
-					Status: metav1.ConditionUnknown,
-				},
-			}
-		}
+	inv := &userv1.Invitation{}
+	if err := s.client.Get(ctx, client.ObjectKey{Name: name}, inv); err != nil {
+		return nil, fmt.Errorf("failed to get invitation: %w", err)
+	}
+
+	tokenValid := inv.Status.Token != "" && inv.Status.ValidUntil.After(time.Now())
+	if inv.IsRedeemed() || !tokenValid || inv.Status.Token != token {
+		l.Info("invalid token")
```

Comment:
Log message is somewhat misleading, even though it's technically correct.

I'd probably go for something like the following at least

```suggestion
		l.Info("invitation already redeemed or invalid token")
```

Even nicer would be that we log why the token is invalid, but that probably needs a bit of extra logic to format the various invalid states for the log message